x/trigger/types/codec.go

```diff
@@ -1,19 +1,16 @@
 package types
 
 import (
-	"github.com/cosmos/cosmos-sdk/codec"
 	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
 )
 
-// ignoring RegisterLegacyAminoCodec registers all the necessary types and interfaces for the
-// double check
+// RegisterInterfaces registers concrete implementations for this module.
 func RegisterInterfaces(registry cdctypes.InterfaceRegistry) {
-	registry.RegisterImplementations(
-		(*sdk.Msg)(nil),
-		&MsgCreateTriggerRequest{},
-		&MsgDestroyTriggerRequest{},
-	)
+	messages := make([]proto.Message, len(AllRequestMsgs))
```

Comment:
Do we need an extra copy/heap allocation for this? RegisterImplementations is only reading, and this would also allow us to get rid of the new import.

`registry.RegisterImplementations((*sdk.Msg)(nil), AllRequestMsgs...)`

Reply:
Yeah.  `RegisterImplementations` takes in a `[]proto.Message`, but the `AllRequestMsgs` variable is a `[]sdk.Msg`. Since `proto.Message` is a subset of `sdk.Msg`, Golang can automatically cast a single entry, but it can't cast slices like that, so we have to copy them.

In previous versions of the SDK, the `sdk.Msg` interface was a `proto.Message` with several other methods. It was important to have some sort of type assertion on them to make sure all required methods were defined. That's why the `AllRequestMsgs` variables are all `[]sdk.Msg`; having them in that list gave us the desired type assertion for free.

Over the various versions, though, the SDK has removed methods from the `sdk.Msg` interface, and it is now just an alternative name for `proto.Message`. If we changed `AllRequestMsgs` to a `[]proto.Message`, we wouldn't have to do this copying. I'm not quite sure what other ramifications that might have, though. But I'd rather leave them as `[]sdk.Msg` since that's the type we're registering them as in these codecs.

Reply:
I actually got this mixed up with `attribute/types/codec.go` where `allRequestMsgs` already has the []proto.Message type.